### ISSUE TITLE
v0.27.0 destroys the performance of our app

### ISSUE THREAD

**Issue by jgb** [DataDog/dd-trace-py#992]
Hello,

when we upgraded from v0.26.0 to v0.27.0 the performance of our Django + Mariadb + elasticsearch application went down.
Requests via the web interface took 3 times as long, often causing gateway timeouts.
Reverting to v0.26.0 of ddtrace resolves the issue.

Any idea on what might be going on here?

Greetings,

Jan

**Comment by jd** [comment]
Not from the top of my head. There were a few changes in 0.27.0 but it's hard to see which one would cause that. If there's any chance you can bisect which particular commit introduces this regression that would help a lot. Or even basic profiling of a request that took too long would give us a few hints.

**Comment by brettlangdon** [comment]
@jgb can you send me an email from your work email, brett.langdon@datadoghq.com

I'd like to gather some more details from you, but they are specific to your organization so we probably don't want to discuss here :)

**Comment by benjaminrigaud** [comment]
Hello everybody :)

I reproduced the issue in local and it is most likely related to 0.27.0

The first difference is we get an error in 0.26.0:

```
ERROR failed_to_send traces to Datadog Agent: HTTP error status 413, reason Request Entity Too Large, message Content-Type: text/plain; charset=utf-8
X-Content-Type-Options: nosniff
Date: Mon, 22 Jul 2019 09:00:02 GMT
Content-Length: 18
```

I assume this big request is now being accepted in 0.27 and is causing the slow down.

Could be related to https://github.com/DataDog/dd-trace-py/commit/40c0eb8ee9a490b7c1cd304333ecffd60700ace2

Still digging :male_detective:

**Comment by jd** [comment]
Yeah that's something we fixed in 0.27.0. I imagine that if the payload is now accepted with 0.27..0, your local agent might be doing more processing. Did you check its CPU usage, just in case?

**Comment by benjaminrigaud** [comment]
It looks like the agent is doing fine (no observable CPU usage yet).
The client (a Django app) is using one core at 100% for the duration of the request (150 seconds in 0.27.0).

**Comment by benjaminrigaud** [comment]
It looks like 0.27.0 is doing a lot in context.py (close_span & get)
![image](https://user-images.githubusercontent.com/1983087/61629003-0de0af00-ac84-11e9-80f1-6348fc981c78.png)

**Comment by jd** [comment]
This could be due to 0252795bdfd1ee23c6fd0a6eb2e1a520f87c665c. Do you have the same profiling with 0.26.0?

**Comment by benjaminrigaud** [comment]
@jd Can you give me your email?
I'll send you both .svg (and .prof if you need) files.

**Comment by jd** [comment]
@benjaminrigaud `jd@datadoghq.com`

I've pushed a piece of code in #996 that might help if https://github.com/DataDog/dd-trace-py/commit/0252795bdfd1ee23c6fd0a6eb2e1a520f87c665c is the culprit. You might want to test it if you can.

**Comment by benjaminrigaud** [comment]
I just tried it but the performance are similar (tracer.debug_logging is False).

Sending the prof / svg by email in a few minutes.

**Comment by brettlangdon** [comment]
v0.27.1 has been released with #996 which resolves this issue.
